services/route_manager/src/main/java/com/futurewei/alcor/route/service/Impl/RouterServiceImpl.java

```diff
@@ -0,0 +1,316 @@
+        this.routeEntryDatabaseService.addRouteEntry(routeEntry);
+
+        RouteTable routeTable = new RouteTable(projectId, routeTableId, "default_vpc_routeTable", "", routeEntities, RouteTableType.VPC, owner);
+        vpcRouteTables.add(routeTable);
+        this.routeTableDatabaseService.addRouteTable(routeTable);
+
+        vpcRouteTables.add(routeTable);
+        router.setVpcRouteTable(vpcRouteTables);
+        this.routerDatabaseService.addRouter(router);
+
+        return routeTable;
+    }
+
+    @Override
+    public RouteTable updateVpcRouteTable(String projectId, String vpcId, RouteTableWebJson resource) throws DatabasePersistenceException, CanNotFindVpc, CacheException, ExistMultipleVpcRouter {
+        RouteTable routeTable = null;
+        RouteTable inRoutetable = resource.getRoutetable();
+
+        // Get or create a router for a Vpc
+        Router router = getOrCreateVpcRouter(projectId, vpcId);
+
+        // check if there is a vpc default routetable
+        List<RouteTable> vpcRouteTables = router.getVpcRouteTable();
+        for (RouteTable vpcRouteTable : vpcRouteTables) {
+            String routeTableType = vpcRouteTable.getRouteTableType().getRouteTableType();
+            if (RouteTableType.VPC.getRouteTableType().equals(routeTableType)) {
+                routeTable = vpcRouteTable;
+                vpcRouteTables.remove(vpcRouteTable);
+                break;
+            }
+        }
+
+        if (routeTable == null) {
+            String routeTableId = inRoutetable.getId();
+            if (routeTableId == null) {
+                routeTableId = UUID.randomUUID().toString();
+                inRoutetable.setId(routeTableId);
+            }
+            inRoutetable.setRouteTableType(RouteTableType.VPC);
+            vpcRouteTables.add(inRoutetable);
+            router.setVpcRouteTable(vpcRouteTables);
+            this.routerDatabaseService.addRouter(router);
+
+            return inRoutetable;
+        } else {
+            List<RouteEntry> routeEntities = routeTable.getRouteEntities();
+            List<RouteEntry> inRouteEntities = inRoutetable.getRouteEntities();
+
+            for (RouteEntry routeEntry : routeEntities) {
+                if (!inRouteEntities.contains(routeEntry)) {
+                    // TODO: check if existing rules are currently used by other subnet’s routing table
+                }
+            }
+            routeTable.setRouteEntities(inRouteEntities);
+            vpcRouteTables.add(routeTable);
+            router.setVpcRouteTable(vpcRouteTables);
+            this.routerDatabaseService.addRouter(router);
+
+            return routeTable;
+        }
+
+    }
+
+    @Override
+    public List<RouteTable> getVpcRouteTables(String projectId, String vpcId) throws CanNotFindVpc {
+        VpcWebJson vpcResponse = this.vpcRouterToVpcService.getVpcWebJson(projectId, vpcId);
+        VpcEntity vpcEntity = vpcResponse.getNetwork();
+        Router router = vpcEntity.getRouter();
```

Comment:
This function should return all routing tables within this VPC, including VPC default routing table and all subnets' routing tables, not just only VPC default route table.

Reply:
What is returned here is a List